data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestMapper.java

```diff
@@ -193,14 +217,61 @@ private static String encoding(final DataPlaneContract.ContentMode contentMode)
     };
   }
 
-  private static class Producer<K, V> {
+  private static class ReferenceCounter<T> {
+    final T value;
+    int count;
+
+    public ReferenceCounter(T value) {
+      this.value = value;
+      this.count = 0;
```

Comment:
If we create it, we have a reference. Should we start at 1 and remove the increment right below the instantiation?

Reply:
IMO this is wrong, because `ReferenceCounter` is the object that has "ownership" of the value: the reference counter could be created, but no refs could be there. Also, it makes the usage nicer this way

Reply:
I'm not sure I understand what you mean by "ownership" in this specific case.

Reply:
I mean that `ReferenceCounter` is in this context "distinct" from who refers to `value` (in this case ingress info) and creating a `ReferenceCounter` doesn't imply creating a reference to `IngressInfo`

Reply:
There is no point _in this context_ in creating a reference counter _with a value_, and the count is 0, in fact, we need to call increment after creating it otherwise the state is not correct by construction. Anyway, let's keep it that way.